### ISSUE TITLE
merge_features

### ISSUE THREAD

**Issue by hongminglin08** [JacobChalk/TIM#28]
When I run merge_features.py, the following error occurs. The feature dimensions extracted by videomae and omnivore seem to be inconsistent, is there any step missing? Here, videomae takes the features directly from your pre-trained model and runs Feature_extrusion.py.
![image](https://github.com/JacobChalk/TIM/assets/52478955/11078dd8-2f51-4559-a1d6-55eae30e62b8)

**Comment by JacobChalk** [comment]
It seems the number of feature sets for your VideoMAE is less than that of Omnivore (4 vs 1). The solution here would be either to drop the last 3 omnivore feature sets (note this will reduce augmentation and negatively impact performance). Or to extract 3 additional feature sets for VideoMAE (note these additional sets should be augmented as TIM operates under the assumption that `[:, 0, :]` are the unaugmented features and `[:, 1:, :]` are the augmented ones).

Out of interest, did you change the `--num_aug` argument when extracting the VideoMAE features [here](https://github.com/JacobChalk/TIM/blob/0ef561e61a8d6c4351f03ec7d680db85a1f515e3/feature_extractors/VideoMAE/feature_extraction.py#L52)? As the code should stack all extracted feature sets as seen [here](https://github.com/JacobChalk/TIM/blob/0ef561e61a8d6c4351f03ec7d680db85a1f515e3/feature_extractors/VideoMAE/feature_extraction.py#L242).

**Comment by hongminglin08** [comment]
> It seems the number of feature sets for your VideoMAE is less than that of Omnivore (4 vs 1). The solution here would be either to drop the last 3 omnivore feature sets (note this will reduce augmentation and negatively impact performance). Or to extract 3 additional feature sets for VideoMAE (note these additional sets should be augmented as TIM operates under the assumption that `[:, 0, :]` are the unaugmented features and `[:, 1:, :]` are the augmented ones).
> 
> Out of interest, did you change the `--num_aug` argument when extracting the VideoMAE features [here](https://github.com/JacobChalk/TIM/blob/0ef561e61a8d6c4351f03ec7d680db85a1f515e3/feature_extractors/VideoMAE/feature_extraction.py#L52)? As the code should stack all extracted feature sets as seen [here](https://github.com/JacobChalk/TIM/blob/0ef561e61a8d6c4351f03ec7d680db85a1f515e3/feature_extractors/VideoMAE/feature_extraction.py#L242).

I didn't change the num_aug parameter. May I ask which scheme you adopted to extract dimensionally consistent features?

**Comment by JacobChalk** [comment]
We ensure that we extract the same number of feature sets for each backbone, in our case this is 4. The first feature set is extracted from an unaugmented frame input, the final 3 feature sets are extracted applying RandAugment on the input frames. This means our dimensions for both backbones are of size `[num_feats, 4, 1024]` where `num_feats` varies depending on the input video.

To fix your issue, and also match our scheme, you will need to extracted 3 additional feature sets of VideoMAE (be careful of the `enable_augmentation` flag on the dataloader [here](https://github.com/JacobChalk/TIM/blob/5016bba14a1d252f15f97cbfa26bc23ca42cd2ec/feature_extractors/VideoMAE/feature_extraction.py#L223)) and concatenate them to the axis of size 1 in the VideoMAE features.

**Comment by hongminglin08** [comment]
> We ensure that we extract the same number of feature sets for each backbone, in our case this is 4. The first feature set is extracted from an unaugmented frame input, the final 3 feature sets are extracted applying RandAugment on the input frames. This means our dimensions for both backbones are of size `[num_feats, 4, 1024]` where `num_feats` varies depending on the input video.
> 
> To fix your issue, and also match our scheme, you will need to extracted 3 additional feature sets of VideoMAE (be careful of the `enable_augmentation` flag on the dataloader [here](https://github.com/JacobChalk/TIM/blob/5016bba14a1d252f15f97cbfa26bc23ca42cd2ec/feature_extractors/VideoMAE/feature_extraction.py#L223)) and concatenate them to the axis of size 1 in the VideoMAE features.

We retrained videomae's python -u feature_extraction.py and found that VisionTransformer lacks two parameters, pretrained_cfg and pretrained_cfg_overlay. Are both parameters set to false?

**Comment by JacobChalk** [comment]
Assuming you mean the VisionTransformer [here](https://github.com/JacobChalk/TIM/blob/51df464c9853b0ce7e8a67f7030b5cc9c2fe381e/feature_extractors/VideoMAE/modeling_finetune.py#L244), I'm unable to see those arguments, so the assumption is that they are indeed false, though we did not seem to set those parameters ourselves.

**Comment by hongminglin08** [comment]
> Assuming you mean the VisionTransformer [here](https://github.com/JacobChalk/TIM/blob/51df464c9853b0ce7e8a67f7030b5cc9c2fe381e/feature_extractors/VideoMAE/modeling_finetune.py#L244), I'm unable to see those arguments, so the assumption is that they are indeed false, though we did not seem to set those parameters ourselves.

It is the num_aug parameter setting. args.num_aug = min(args.num_aug, 1) in line 70 of feature_extraction.py causes the parameter to be 1. This line should be commented.

**Comment by JacobChalk** [comment]
Good spot! Thank you for highlighting this issue, it has now been updated to the correct line and should now work as intended.
